optaplanner-core/src/main/java/org/optaplanner/core/impl/score/stream/drools/common/rules/TriExistenceMutator.java

```diff
@@ -0,0 +1,102 @@
+        }
+        // There is no delta index in Drools, therefore we replace joining with a filter.
+        QuadPredicate<A, B, C, D> joinFilter = joiner::matches;
+        QuadPredicate<A, B, C, D> result = predicate == null ? joinFilter : joinFilter.and(predicate);
+        // And finally we add the filter to the D pattern,
+        return applyFilters(ruleAssembler, result);
+    }
+
+    private AbstractRuleAssembler applyFilters(AbstractRuleAssembler ruleAssembler, QuadPredicate<A, B, C, D> predicate) {
+        Variable<D> toExist = PatternDSL.declarationOf(otherFactType, ruleAssembler.generateNextId("triToExist"));
+        PatternDSL.PatternDef<D> existencePattern = PatternDSL.pattern(toExist);
+        PatternDSL.PatternDef<D> possiblyFilteredExistencePattern = predicate == null ? existencePattern
+                : existencePattern.expr("Filter using " + predicate, ruleAssembler.getVariable(0),
+                        ruleAssembler.getVariable(1), ruleAssembler.getVariable(2),
+                        (d, a, b, c) -> predicate.test((A) a, (B) b, (C) c, d));
+        ExprViewItem existenceExpression = PatternDSL.exists(possiblyFilteredExistencePattern);
+        if (!shouldExist) {
+            existenceExpression = PatternDSL.not(possiblyFilteredExistencePattern);
+        }
+        ruleAssembler.addDependentExpressionToLastPattern(existenceExpression);
+        return ruleAssembler;
+    }
+
+    @Override
+    public AbstractRuleAssembler apply(AbstractRuleAssembler ruleAssembler) {
+        int indexOfFirstFilter = -1;
+        // Prepare the joiner and filter that will be used in the pattern
+        AbstractQuadJoiner<A, B, C, D> finalJoiner = null;
+        QuadPredicate<A, B, C, D> finalFilter = null;
+        for (int i = 0; i < joiners.length; i++) {
+            AbstractQuadJoiner<A, B, C, D> joiner = joiners[i];
+            boolean hasAFilter = indexOfFirstFilter >= 0;
+            if (joiner instanceof NoneQuadJoiner && joiners.length > 1) {
+                throw new IllegalStateException("If present, " + NoneTriJoiner.class + " must be the only joiner, got "
+                        + Arrays.toString(joiners) + " instead.");
+            } else if (!(joiner instanceof FilteringQuadJoiner)) {
+                if (hasAFilter) {
+                    throw new IllegalStateException("Indexing joiner (" + joiner + ") must not follow a filtering joiner ("
+                            + joiners[indexOfFirstFilter] + ").");
+                } else { // Merge this Joiner with the existing Joiners.
+                    finalJoiner = finalJoiner == null ? joiner : AbstractQuadJoiner.merge(finalJoiner, joiner);
+                }
+            } else {
+                if (!hasAFilter) { // From now on, we only allow filtering joiners.
+                    indexOfFirstFilter = i;
+                }
+                // We merge all filters into one, so that we don't pay the penalty for lack of indexing more than once.
```

Comment:
comment code style: avoid "we" and "you" just like docs -> "Merge all filters into on to avoid paying the penalty for ..."

Motivation: by always writing in the same style, it's easier to switch back and forth between docs, blogs and comments. It's also a better writting style.